src/kdl/serializable.cr

```diff
@@ -0,0 +1,361 @@
+module KDL
+  annotation Argument; end
+  annotation Arguments; end
+  annotation Property; end
+  annotation Properties; end
+  annotation Child; end
+  annotation Children; end
+
+  module Serializable
+    class Error < Exception
+      getter klass : String
+      getter attribute : String?
+
+      def initialize(message : String?, @klass : String, @attribute : String?)
+        message = String.build do |io|
+          io << message
+          io << "\n  parsing "
+          io << klass
+          if attribute = @attribute
+            io << '#' << attribute
+          end
+        end
+        super(message)
+      end
+    end
+
+    macro convert(expr, type)
+      {% type_str = type.stringify %}
+      {% if type_str == "Int8" %}
+        {{expr}}.as(Number).to_i8
+      {% elsif type_str == "Int16" %}
+        {{expr}}.as(Number).to_i16
+      {% elsif type_str == "Int32" %}
+        {{expr}}.as(Number).to_i32
+      {% elsif type_str == "Int64" %}
+        {{expr}}.as(Number).to_i64
+      {% elsif type_str == "Int128" %}
+        {{expr}}.as(Number).to_i128
+      {% elsif type_str == "UInt8" %}
+        {{expr}}.as(Number).to_u8
+      {% elsif type_str == "UInt16" %}
+        {{expr}}.as(Number).to_u16
+      {% elsif type_str == "UInt32" %}
+        {{expr}}.as(Number).to_u32
+      {% elsif type_str == "UInt64" %}
+        {{expr}}.as(Number).to_u64
+      {% elsif type_str == "UInt128" %}
+        {{expr}}.as(Number).to_u128
+      {% elsif type_str == "Float32" %}
+        {{expr}}.as(Number).to_f32
+      {% elsif type_str == "Float64" %}
+        {{expr}}.as(Number).to_f64
+      {% else %}
+        {{expr}}.as({{type}})
+      {% end %}
+    end
+  
+    macro included
+      def self.from_kdl(doc)
+        new doc
+      end
+    
+      def self.new(node : ::KDL::Node)
+        new_from_kdl_node(node)
+      end
+
+      def self.new(doc : ::KDL::Document)
+        new_from_kdl_node(::KDL::Node.new("__root", children: doc))
+      end
```

Comment:
An additional initialize method would be needed to allow newing up fully defined classes / structs.
```cr
def initialize
end
```
Otherwise something like
```cr
  class Mappings
    include KDL::Serializable

    @[KDL::Property(name: "update-frequency")]
    property update_frequency = 1

    @[KDL::Property(name: "last-update")]
    property last_update = 0
  end
  ```
  will complain that `Mappings.new` needs an argument.

Reply:
Actually this breaks the spec, so will have to figure something else out :thinking:

Reply:
Yeah, we can't have a default initializer unless all properties have a default value. Alternatively, we ditch the `new` and just do everything through `from_kdl`

Reply:
Any way would work for me :D Currently I'm manually defining a default initializer in all fully defined classes to make it compile, but an automatic solution would be cool

Reply:
I'll see what I can do. If all ivars have default values I can auto-define a default initializer

Reply:
I was incorrectly under the assumption that `::JSON::Serializable` did it automatically. [It does not](https://github.com/crystal-lang/crystal/blob/dacd97bcc/src/json/serialization.cr#L152) and requires a manual `def initialize; end` as well.

Never mind then :D

I think the PR is ready to be merged in that case. Excellent work! :clap: